[i18n] Add BCP47 Language Tags Parsing

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Implement a BCP 47 language tags parser with full support of RFC5646, a fast path for tags made up of the language subtag only (ex: en, ara), and convert a few code paths to use the new functionality.

### Language descriptions

The English language description displayed for subtitles and audio tracks now supports bcp47 language tags and mimicks locale names. It is however limited by the lack of support of the IANA language subtags registry and for now only the recently updated existing ISO639-1/-2 and ISO 3166-1 support is used to convert languages and regions to their English names.

For example:
- en-AU => English (Australia) (en is ISO 639-1 and AU is ISO 3166-1)
- es-419 => Spanish (419) (es is ISO 639-1, 419 is M.49 and requires the registry)
- zh-yue-Hant-HK => Chinese (yue Hant Hong Kong) (extlang yue and script Hant require the registry)
- yue-HK => yue (Hong Kong) (yue is ISO 639-3 and requires the registry)

As was already happening before the PR, user languages defined in as.xml or language addons take precedence and provide "friendlier" names.

### External subtitles / audio tracks

Modified the parsing of external subtitles /audio tracks to accept bcp47 language tags in the filename. The - separators have to be replaced with _ since - is already a segment separator for filenames. For example:
Big.Buck.Bunny.forced.**en_AU**.srt

### Movies and TV Shows original language

The original language scraped or imported from nfo is now stored as BCP47 in the video db. Existing entries were converted from ISO 639-2/B to BCP47 by a DB bump.

The original language field was missed in a couple places of the VideoInfoTag when it was added and this is now corrected.


### Other support of BCP47 in media

There is nothing to change for InputStream as far as I can tell, it already forwards bcp47 language tags to core and benefits from the improved conversion to English language names of the PR.

Internal tracks of file-based media are problematic due to missing support of BCP47 in ffmpeg, at least for mkv (mp4 seems problematic as well).
The workaround of entering the language tag in the track name, surrounded by curly braces, was upgraded to the new BCP 47 parser.

### Future PR
- a fast path could be added for language+region or language+script if it is felt necessary, in order to avoid the full blown regexp parsing for common/easy tags
- implement the match algorithms of BCP47 to improve the selection of the best audio or subtitle track to
- implement the IANA language subtags registry for better conversion of BCP47 to English and for validation of the tags
- internally store track languages as BCP47 objects instead of string to reduce the amount of parsing
- alternatively implement caching of parsed tags
- what to do about the translation of language names themselves - the registry or ISO norms only contain English names for all subtags

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Many code paths were limited to 3 letter ISO 639-2 codes, not sufficient to express the variety of languages.
BCP 47 is the standard.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Added unit tests for everything. All pass.
Confirmed that external subs, mkv, ia have the additional subtags, and that those are translated to English on screen (within limits imposed by lack of support of the IANA language subtags registry.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

more languages recognized without having to define entries in advancedsettings.xml/language codes or requiring language addons to be installed.

## Screenshots (if appropriate):
external subs
<img width="1114" height="516" alt="image" src="https://github.com/user-attachments/assets/ac1e22f2-78d0-4df4-8f4d-70b6d09b4a66" />

inputstream adaptative
<img width="1107" height="800" alt="image" src="https://github.com/user-attachments/assets/e0a97fb5-61f4-48d1-b23d-cb941493a3fa" />

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

CrystalP committed:
[VideoInfoTag] Add original language to Reset and serialization methods

CrystalP committed:
[VideoDB] DB bump to convert existing ISO 639-2/B codes to BCP 47

CrystalP committed:
[Regexp] Add retrieval of captures by name

CrystalP committed:
[VideoInfoTag] Use BCP 47 for movies/tv shows original language

CrystalP commented:
- Database migration: needed some finetuning - done.
- Parser regexp correctness/edge cases: verified by many unit tests - added more
- The validity of BCP 47 tags cannot be implemented without the registry, which cannot happen in this PR as it would make it a monster. I implemented what could be given the ISO data already available.
-  Integration Points in LangCodeExpander: unit tests for the various inputs - added more

About the questions to the author
- I looked (not very long though...) for an existing library. ICU deals with localization. The planned usage of this code has nothing to do with localization, also locale ids look a lot like BCP47 language tags but they're not the same thing. So it doesn't look like a good fit to me. It would also be another external dependency.
- nothing changes for a large library. The migration is written so that the number of update queries is the minimum possible (ex. update all records with original language "eng" at once)
- perf measurement: I don't know since there is not much real data yet. As indicated in PR description, more fast paths could be added for the lang-region form or a lang-script form which are probably relatively common.
- Code for IANA registry is mostly ready, though integration with code of this PR is not done yet.

CrystalP committed:
[i18n] Remove CBcp47Formatter as friend of CBcp47

The interface offered by CBcp47 was a mix of public accessors and friend access to private members.

The main users of the class haven't been written yet and it has yet to be determined that public accessors
won't make things more complicated by requiring the callers to have knowledge of the internals of CBcp47
(interpret the tag type to access different fields)
For now public accessors make the unit tests easier to write but this may be reversed down the road and
a separate commit will make the reversal easy.

CrystalP commented:
fixed clang-format... & log an error when a tag or description from nfo or scraper is not recognized.

CrystalP committed:
Add BCP47 class/parser

CrystalP committed:
[VideoPlayer] Enable BCP47 language tags for external subs

CrystalP committed:
[LanguageCodes] Adapt mkv/ffmpeg workaround to use the BCP 47 parser

CrystalP committed:
[GUI] Convert language and region subtags to English for display

CastagnaIT commented:
the only thing i found are uncomfortable ISO3166_1 names, are really too long,
example:
<img width="644" height="553" alt="immagine" src="https://github.com/user-attachments/assets/e875b084-67df-446f-91be-eaa33f52cb43" />

and there are even longer names

this will cause considerable confusion in the list, making it difficult to read the tracks
IMO we should consider shortening the names listed (separate PR)

CrystalP commented:
Had the CBcp47 accessors return const references and adapters a few callers that could take advantage of it.

CrystalP commented:
no functional changes:
rebased on master for easier build.
Renamed the new TestUtils.h/c.pp files to TestI18nUtils.h/.cpp for clarity in IDE - there already is a TestUtils..h/cpp pair in xbmc/test and it's more logical for them to "own" the name.

CastagnaIT reviewed (approved):
i dont have any other question, from my side its ok,
the only thing that i find hard to understand is the regex parser,
i tried checking it with an AI and it seems to be sufficient

CrystalP commented:
Yeah the regexp... that was fun,  I was (still am) a regexp newbie, it took some time with an online checker to figure out the basic structure. It covers the full grammar of the RFC except for the grandfathered tags, handled separately.
I added unit tests to prove that it works as expected.
Thanks for the review!

CrystalP commented:
@claude any final comments?

CrystalP commented:
@CastagnaIT I made some progress on the bcp47 registry and updated the PR #27411 if you want to give it a try.

It includes additional commits to start taking advantage of the library.
The tag parser (for .nfo, scrapers) accepts all valid language tags (for example the formerly rejected ISO 639-3 yue for Cantonese and numeric countries like 419 of the common es-419)
Formatting the tags as English (for language info labels) uses the registry for all components.

It helps in the case of en-US where US is now mapped to "United States" instead of "United States of America (the)" but is still not ideal in other cases.
For example yue is mapped to "Yue Chinese", although the registry also has the probably preferable "Cantonese", but how to know when to use the first or the nth description... A simplistic heuristic would be to pick the shortest one - happens to work for yue but no idea for the more general case.
The ISO tables may have some preferable English names... I just don't know enough about the subject. Some finetuning may be required later.

rockrider69 commented:
@CrystalP
more of a question here ;) as I am not sure it is fully related to this PR.
I am maintaining an addon called "Language Preference Manager".

From Piers, I see that external subs are now reported in Json call (Player.GetProperties fields) as 2-digits language codes in (ex. en, fr, ...). It was 3-digits up to now.

The internal subtitles streams (mkv, mp4) are still reported as 3-digits codes, unchanged.

Why this difference ? Are the internal streams going to move to 2-digits as well in the future ?

CrystalP commented:
The implementation is in-between two worlds, I'm not sure I'll be able to finish in v22.
The language of external audio and subs (and inputstream streams) is parsed and returned as BCP47, meaning a 2 letter code when one exists for the language, 3 letter code otherwise. With optional country, script, etc. appended after the language code.

Yes the plan is to move everything to BCP47. With a BCP47-aware comparison function, to match user languages with the stream languages.
ffmpeg is a roadblock though as it cannot read the BCP47 tag from mkv or mp4 for example and retrieves only the backwards compatible iso659.

rockrider69 commented:
@CrystalP 
thx a lot for the answer. Looks like I need to make a plan for progressive BCP47 support as well ;)
